### PR TITLE
updated DataLoader, added a optional keyword argument rng.

### PR DESCRIPTION
**Pull request by norci**
in order to get a reproducible result, without using the global rng.

### CODE REVIEW

File: src/data/dataloader.jl [hunk 1]
```diff
@@ -1,17 +1,18 @@
 # Adapted from Knet's src/data.jl (author: Deniz Yuret)
-
-struct DataLoader{D}
+using Random: AbstractRNG, shuffle!, GLOBAL_RNG
```
Comment (CarloLucibello) [inline src/data/dataloader.jl:2]: add newline below

File: test/data.jl [hunk 2]
```diff
@@ -1,3 +1,4 @@
+using Random
```
Comment (CarloLucibello) [inline test/data.jl:1]: newline below

**Comment by CarloLucibello** [comment]
it seems reasonable to do this

**Comment by DhairyaLGandhi** [comment]
Do we gain much by storing the RNG? Could that be avoided?

**Comment by darsnack** [comment]
I feel like you can't avoid storing it, because it is used in `Base.iterate`. Passing it in as state would be too clunky.

**Comment by DhairyaLGandhi** [comment]
I think I'm wondering whether we are better off supporting custom iterators Instead and have these kinds of details handled by the that function instead

**Comment by CarloLucibello** [comment]
i don't know what you mean, DataLoader is a custom iterator already

**Comment by DhairyaLGandhi** [comment]
Let me do a rough draft of what I have in mind tonight

**Comment by darsnack** [comment]
This seems like a small, simple change that would be nice to get in before the next release. It's breaking right? Would be a shame to wait until 0.13 to use user-specified RNGs.

**Comment by CarloLucibello** [comment]
>  It's breaking right?

doesn't look breaking, just another keyword arg with the default argument which was implicitly used already

**Comment by ToucheSir** [comment]
After thinking about this a bit more, why not direct users to something like `MLDataPattern.shuffleobs` that already takes an RNG argument? `DataLoader` has always been awkward when it comes to extensibility, but I'm not sure the solution is to add more arguments (I'll refrain from rambling about `DataLoader` vs `DataLoaders.DataLoader` here :stuck_out_tongue:).

**Comment by darsnack** [comment]
I think we should hold-off on pushing MLDataPattern.jl too much until it has been refactored (i.e. `MLDataPattern` has a `DataLoader` that can replaces `DataLoaders.DataLoader`). This is fairly simple and easy feature add.

PS: wow that sentence was a mouthful

File: test/data.jl [hunk 3]
```diff
@@ -74,4 +76,7 @@
     d  = DataLoader((X, Y))
     Flux.train!(loss, [θ], ncycle(d, 10), Descent(0.1))
     @test norm(θ .- 1) < 1e-10
+
+    # specify the rng
+    d = map(identity, DataLoader(X, batchsize=2; shuffle=true, rng=Random.seed!(Random.default_rng(), 5)))
```
Comment (DhairyaLGandhi) [inline test/data.jl:81]: We would need better tests to ensure that the expected behaviour is happening. That the API works is pretty good, but we need to ensure that the functionality works as expected.

**Comment by DhairyaLGandhi** [comment]
I would want the whole DataLoader to be extensible, so we can use it for things like training GANs/ NODEs, tuples of arguments (unlike just working on arrays), sending data to different workers/ devices asynchronously etc. It would be cool to have a package to take care of the heavy lifting here, but failing that, I would want these features before moving ahead with MLData* packages.

This should be its own issue by the looks of it.

**Review by CarloLucibello — APPROVED** [review]

**Comment by DhairyaLGandhi** [comment]
@CarloLucibello there's pending comments about lacking tests and this is an api change. Further, regardless of merging this or #1508, the comments need to be addressed. We were not in a rush for merging this, anyway.
